rust/kernel/virtio.rs

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Virtio.
+//!
+//! C header: [`include/linux/virtio.h`](../../../../include/linux/virtio.h)
+
+use crate::{
+    bindings, device, driver, error::from_kernel_result, str::CStr, to_result,
+    types::PointerWrapper, Result, ThisModule,
+};
+
+/// A registration of a virtio driver.
+pub type Registration<T> = driver::Registration<Adapter<T>>;
+
+/// Id of an Virtio device.
+#[derive(Clone, Copy)]
+pub struct DeviceId {
+    /// Device id.
+    pub device: u32,
+
+    /// Vendor id of the virtio device.
+    pub vendor: u32,
+}
+
+// SAFETY: `ZERO` is all zeroed-out and `to_rawid` stores in `virtio_id::data`.
+unsafe impl const driver::RawDeviceId for DeviceId {
+    type RawType = bindings::virtio_device_id;
+    const ZERO: Self::RawType = bindings::virtio_device_id {
+        device: 0,
+        vendor: 0,
+    };
+
+    fn to_rawid(&self, _offset: isize) -> Self::RawType {
+        bindings::virtio_device_id {
+            device: self.device,
+            vendor: self.vendor,
+        }
+    }
+}
+
+/// An adapter for the registration of virtio drivers.
+pub struct Adapter<T: Driver>(T);
+
+impl<T: Driver> driver::DriverOps for Adapter<T> {
+    type RegType = bindings::virtio_driver;
+
+    unsafe fn register(
+        reg: *mut bindings::virtio_driver,
+        name: &'static CStr,
+        module: &'static ThisModule,
+    ) -> Result {
+        // SAFETY: By the safety requirements of this function (defined in the trait definition),
+        // `reg` is non-null and valid.
+        let vdrv = unsafe { &mut *reg };
+
+        vdrv.driver.name = name.as_char_ptr();
+        vdrv.driver.owner = module.0;
+        vdrv.probe = Some(Self::probe_callback);
+        vdrv.remove = Some(Self::remove_callback);
+        if let Some(t) = T::ID_TABLE {
+            vdrv.id_table = t.as_ref();
+        }
```

Comment:
If you want `of` ids to work here, you must set `vdrv.driver.of_match_table` -- see `platform.rs` for an example.

Reply:
I found that in the virtio drivers of C sides, the common way to match the driver with the device is to use 'id_table' rather than `of`. I write a macro `define_virtio_id_table` to implement this. It's more like `define_amba_id_table` in the `amba.rs`.